pkg/api/v1beta1/types.go

```diff
@@ -397,6 +397,23 @@ type EnvVar struct {
 	Key  string `json:"key,omitempty" description:"name of the environment variable; must be a C_IDENTIFIER; deprecated - use name instead"`
 	// Optional: defaults to "".
 	Value string `json:"value,omitempty" description:"value of the environment variable; defaults to empty string"`
+	// Optional: specify a source for the value of the EnvVar.
+	Source *EnvVarSource `json:"source,omitempty" description:"source for the environment variable's value; cannot be used if value is not empty"`
+}
+
+// EnvVarSource represents a source for the value of an EnvVar.
+// Only one of its members may be specified.
+type EnvVarSource struct {
+	// Selects a field of the pod; only name and namespace are supported.
+	PodField *ObjectFieldSelector `json:"podField,omitempty" description:"selectes a field of the pod; only name and namespace are supported"`
+}
+
+// ObjectFieldSelector selects an APIVersioned field of an object.
+type ObjectFieldSelector struct {
+	// The API version the FieldPath is written in terms of.
+	APIVersion string `json:"apiVersion,omitempty" description="The API version that FieldPath is written in terms of"`
```

Comment:
If you want to use ObjectFieldSelector, document that it must be a strict subset of ObjectReference.

My bigger beef is with the field name: podField. I'm 99% certain that we're going to need to be able to substitute fields of other objects. In the case that we support that in the API, podField is awkward. In the case that we push cross-object substitution to the client, podField seems both over- and under-specified (of course it must be a pod, but it can't be another pod).

I'll try to provide a more constructive suggestion later tonight.